include/deal.II/lac/sparse_matrix.h

```diff
@@ -2377,6 +2384,35 @@ SparseMatrix<number>::end (const size_type r)
 }
 
 
+template <typename number>
+template <class StreamType>
+void SparseMatrix<number>::print_matrix_market(StreamType &out,
+					       const number threshold) const
+{
+  Assert(cols != 0, ExcNotInitialized());
+  Assert(val != 0, ExcNotInitialized());
+  Assert( threshold > 0, dealii::ExcMessage("Negative threshold!") ); 
+
+  //Print the header
+  out << "%%MatrixMarket matrix coordinate real general\n";
```

Comment:
Yes, I think that's exactly what's needed. Having a free function also makes it much easier to do partial specializations of the scalar type, in exactly the way you're doing here.